state-chain/pallets/cf-ingress-egress/src/lib.rs

```diff
@@ -1179,6 +1183,9 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 				(available_amount, Zero::zero())
 			});
 
+		// XXX: it is correct not to put the fee if the input amount is insufficient
+		// XXX: but the remaining amount in this case should be zero — so that no free egress would
+		// be performed
```

Comment:
I think it is correct to take the fee even if remaining is zero. The network needs to take the required fee (or as much as possible) in all cases to protect the network against DOS attacks.